backend/api/Services/DocumentService.cs

```diff
@@ -129,29 +134,88 @@ public async Task<DocumentUploadResponse> UploadActivityDocumentAsync(long activ
 
-            // update the pims document status
-            PimsDocument existingDocument = documentRepository.Get(documentId);
-            if (existingDocument is not null)
+            DocumentUpdateResponse response = new DocumentUpdateResponse()
             {
-                existingDocument.DocumentStatusTypeCode = updateRequest.DocumentStatusCode;
-                documentRepository.Update(existingDocument);
+                MetadataExternalResult = new List<ExternalResult<DocumentMetadata>>(),
+            };
+
+
+            var metadataUpdateSucessfull = false;
+            if (updateRequest.DocumentMetadata.Count > 0)
+            {
+                // Retrieve the existing metadata and check if it needs to be updated.
+                ExternalResult<QueryResult<DocumentMetadata>> existingMetadata = await documentStorageRepository.GetDocumentMetadataAsync(updateRequest.MayanDocumentId);
+
+                List<DocumentMetadataUpdateModel> updates = new List<DocumentMetadataUpdateModel>();
+                List<DocumentMetadataUpdateModel> creates = new List<DocumentMetadataUpdateModel>();
+                List<DocumentMetadataUpdateModel> deletes = new List<DocumentMetadataUpdateModel>();
+
+                foreach (var updateMetadata in updateRequest.DocumentMetadata)
+                {
+                    var existing = existingMetadata.Payload.Results.FirstOrDefault(x => x.MetadataType.Id == updateMetadata.MetadataTypeId);
+
+                    if (existing == null)
+                    {
+                        creates.Add(updateMetadata);
+                    }
+                    else if (existing.Value != updateMetadata.Value && !string.IsNullOrEmpty(updateMetadata.Value))
+                    {
+                        updateMetadata.Id = existing.Id;
+                        updates.Add(updateMetadata);
+                    }
+                    else if (string.IsNullOrEmpty(updateMetadata.Value))
+                    {
+                        updateMetadata.Id = existing.Id;
+                        deletes.Add(updateMetadata);
+                    }
+                }
+
+                // Update metadata of document
+                response.MetadataExternalResult.AddRange(await UpdateMetadata(updateRequest.MayanDocumentId, updates));
+
+                // Create metadata of document
+                response.MetadataExternalResult.AddRange(await CreateMetadata(updateRequest.MayanDocumentId, creates));
+
+                // Delete metadata of document
+                response.MetadataExternalResult.AddRange(await DeleteMetadata(updateRequest.MayanDocumentId, deletes));
+
+                // Add the metadata response
+                foreach (var task in response.MetadataExternalResult)
+                {
+                    // Flag to know if at least one call was successful.
+                    metadataUpdateSucessfull = metadataUpdateSucessfull || task.Status == ExternalResultStatus.Success;
```

Comment:
hmm, aren't we more concerned if one or more failed to report that to the user? I'd rather get an error if I had 99 successes and one failures just so that I know something didn't complete as expected.

Reply:
We actually do send every outcome status to the front end, is just that for the purpose of updating the DB, if at least one is successful we proceed with updating our side. We can switch it around to be an all or none case if necessary